// knex-schema-inspector/dist/dialects/oracledb.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rawColumnToColumn = void 0;
function rawColumnToColumn(rawColumn) {
    return {
        name: rawColumn.COLUMN_NAME,
        table: rawColumn.TABLE_NAME,
        data_type: rawColumn.DATA_TYPE,
        default_value: rawColumn.DATA_DEFAULT,
        max_length: rawColumn.DATA_LENGTH,
        numeric_precision: rawColumn.DATA_PRECISION,
        numeric_scale: rawColumn.DATA_SCALE,
        is_generated: rawColumn.VIRTUAL_COLUMN === 'YES',
        is_nullable: rawColumn.NULLABLE === 'Y',
        is_unique: rawColumn.CONSTRAINT_TYPE === 'U',
        is_primary_key: rawColumn.CONSTRAINT_TYPE === 'P',
        has_auto_increment: rawColumn.IDENTITY_COLUMN === 'YES',
        foreign_key_column: rawColumn.REFERENCED_COLUMN_NAME,
        foreign_key_table: rawColumn.REFERENCED_TABLE_NAME,
        comment: rawColumn.COLUMN_COMMENT,
    };
}
exports.rawColumnToColumn = rawColumnToColumn;
var oracleDB = /** @class */ (function () {
    function oracleDB(knex) {
        this.knex = knex;
    }
    // Tables
    // ===============================================================================================
    /**
     * List all existing tables in the current schema/database
     */
    oracleDB.prototype.tables = function () {
        return __awaiter(this, void 0, void 0, function () {
            var records;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.knex
                            .select('TABLE_NAME')
                            .from('USER_TABLES')];
                    case 1:
                        records = _a.sent();
                        return [2 /*return*/, records.map(function (_a) {
                                var TABLE_NAME = _a.TABLE_NAME;
                                return TABLE_NAME;
                            })];
                }
            });
        });
    };
    oracleDB.prototype.tableInfo = function (table) {
        return __awaiter(this, void 0, void 0, function () {
            var query, rawTable, records;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        query = this.knex.select('TABLE_NAME').from('USER_TABLES');
                        if (!table) return [3 /*break*/, 2];
                        return [4 /*yield*/, query
                                .andWhere({ TABLE_NAME: table })
                                .first()];
                    case 1:
                        rawTable = _a.sent();
                        return [2 /*return*/, {
                                name: rawTable.TABLE_NAME,
                            }];
                    case 2: return [4 /*yield*/, query];
                    case 3:
                        records = _a.sent();
                        return [2 /*return*/, records.map(function (rawTable) {
                                return {
                                    name: rawTable.TABLE_NAME,
                                };
                            })];
                }
            });
        });
    };
    /**
     * Check if a table exists in the current schema/database
     */
    oracleDB.prototype.hasTable = function (table) {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.knex
                            .count({ count: '*' })
                            .from('USER_TABLES')
                            .where({ TABLE_NAME: table })
                            .first()];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, (result && result.count === 1) || false];
                }
            });
        });
    };
    // Columns
    // ===============================================================================================
    /**
     * Get all the available columns in the current schema/database. Can be filtered to a specific table
     */
    oracleDB.prototype.columns = function (table) {
        return __awaiter(this, void 0, void 0, function () {
            var query, records;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        query = this.knex
                            .select('TABLE_NAME', 'COLUMN_NAME')
                            .from('USER_TAB_COLS');
                        if (table) {
                            query.where({ TABLE_NAME: table });
                        }
                        return [4 /*yield*/, query];
                    case 1:
                        records = _a.sent();
                        return [2 /*return*/, records.map(function (_a) {
                                var TABLE_NAME = _a.TABLE_NAME, COLUMN_NAME = _a.COLUMN_NAME;
                                return ({
                                    table: TABLE_NAME,
                                    column: COLUMN_NAME,
                                });
                            })];
                }
            });
        });
    };
    oracleDB.prototype.columnInfo = function (table, column) {
        return __awaiter(this, void 0, void 0, function () {
            var query, rawColumn, records;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        query = this.knex
                            .select('c.TABLE_NAME', 'c.COLUMN_NAME', 'c.DATA_DEFAULT', 'c.DATA_TYPE', 'c.DATA_LENGTH', 'c.DATA_PRECISION', 'c.DATA_SCALE', 'c.NULLABLE', 'c.IDENTITY_COLUMN', 'c.VIRTUAL_COLUMN', 'cm.COMMENTS as COLUMN_COMMENT', 'pk.CONSTRAINT_TYPE', 'fk.REFERENCED_TABLE_NAME', 'fk.REFERENCED_COLUMN_NAME')
                            .from('USER_TAB_COLS as c')
                            .leftJoin('USER_COL_COMMENTS as cm', {
                            'c.TABLE_NAME': 'cm.TABLE_NAME',
                            'c.COLUMN_NAME': 'cm.COLUMN_NAME',
                        })
                            .leftJoin(this.knex
                            .select('uc.CONSTRAINT_TYPE', 'uc.TABLE_NAME', 'cc.COLUMN_NAME')
                            .from('USER_CONSTRAINTS as uc')
                            .join('USER_CONS_COLUMNS as cc', 'uc.CONSTRAINT_NAME', 'cc.CONSTRAINT_NAME')
                            .whereIn('uc.CONSTRAINT_TYPE', ['P', 'U'])
                            .as('pk'), {
                            'c.TABLE_NAME': 'pk.TABLE_NAME',
                            'c.COLUMN_NAME': 'pk.COLUMN_NAME',
                        })
                            .leftJoin(this.knex
                            .select('uc.TABLE_NAME', 'cc.COLUMN_NAME', 'rc.TABLE_NAME as REFERENCED_TABLE_NAME', 'rc.COLUMN_NAME as REFERENCED_COLUMN_NAME')
                            .from('USER_CONSTRAINTS as uc')
                            .as('uc')
                            .join('USER_CONS_COLUMNS as cc', 'uc.CONSTRAINT_NAME', 'cc.CONSTRAINT_NAME')
                            .join('USER_CONS_COLUMNS as rc', 'uc.R_CONSTRAINT_NAME', 'rc.CONSTRAINT_NAME')
                            .where('uc.CONSTRAINT_TYPE', 'R')
                            .as('fk'), {
                            'c.TABLE_NAME': 'fk.TABLE_NAME',
                            'c.COLUMN_NAME': 'fk.COLUMN_NAME',
                        });
                        if (table) {
                            query.where({ 'c.TABLE_NAME': table });
                        }
                        if (!column) return [3 /*break*/, 2];
                        return [4 /*yield*/, query
                                .andWhere({ 'c.COLUMN_NAME': column })
                                .first()];
                    case 1:
                        rawColumn = _a.sent();
                        return [2 /*return*/, rawColumnToColumn(rawColumn)];
                    case 2: return [4 /*yield*/, query];
                    case 3:
                        records = _a.sent();
                        return [2 /*return*/, records.map(rawColumnToColumn)];
                }
            });
        });
    };
    /**
     * Check if a table exists in the current schema/database
     */
    oracleDB.prototype.hasColumn = function (table, column) {
        return __awaiter(this, void 0, void 0, function () {
            var count;
            return __generator(this, function (_a) {
                count = this.knex
                    .count({ count: '*' })
                    .from('USER_TAB_COLS')
                    .where({
                    TABLE_NAME: table,
                    COLUMN_NAME: column,
                })
                    .first().count;
                return [2 /*return*/, !!count];
            });
        });
    };
    /**
     * Get the primary key column for the given table
     */
    oracleDB.prototype.primary = function (table) {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.knex
                            .select('cc.COLUMN_NAME')
                            .from('USER_CONSTRAINTS as uc')
                            .leftJoin('USER_CONS_COLUMNS as cc', 'uc.CONSTRAINT_NAME', 'cc.CONSTRAINT_NAME')
                            .where({
                            'uc.CONSTRAINT_TYPE': 'P',
                            'uc.TABLE_NAME': table,
                        })
                            .first()];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result ? result.COLUMN_NAME : null];
                }
            });
        });
    };
    // Foreign Keys
    // ===============================================================================================
    oracleDB.prototype.foreignKeys = function (table) {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.knex.raw("\n      SELECT\n        ucc. \"TABLE_NAME\" AS \"table\",\n        ucc. \"COLUMN_NAME\" AS \"column\",\n        rucc. \"TABLE_NAME\" AS \"foreign_key_table\",\n        rucc. \"COLUMN_NAME\" AS \"foreign_key_column\",\n        uc. \"CONSTRAINT_NAME\" AS \"constraint_name\",\n        NULL AS \"on_update\",\n        uc. \"DELETE_RULE\" AS \"on_delete\"\n      FROM\n        USER_CONSTRAINTS uc\n        LEFT JOIN USER_CONS_COLUMNS ucc ON uc. \"CONSTRAINT_NAME\" = ucc. \"CONSTRAINT_NAME\"\n        LEFT JOIN USER_CONS_COLUMNS rucc ON uc. \"R_CONSTRAINT_NAME\" = rucc. \"CONSTRAINT_NAME\"\n      WHERE\n        uc. \"CONSTRAINT_TYPE\" = 'R'\n    ")];
                    case 1:
                        result = _a.sent();
                        if (table) {
                            return [2 /*return*/, result === null || result === void 0 ? void 0 : result.filter(function (row) { return row.table === table; })];
                        }
                        return [2 /*return*/, result];
                }
            });
        });
    };
    return oracleDB;
}());
exports.default = oracleDB;
